### PR TITLE
Implement the multiplayer spectator screen

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/12537
- [x] https://github.com/ppy/osu-server-spectator/pull/59

Preview: https://drive.google.com/file/d/1ruPH3Q8z1FHFIvGZK1YNZiyVuVg5Zgcm/view

Quite a big PR, but it should be pretty simple to go through in a linear fashion.

Known issues:
- Spectating players may not show the results screen if spectated mid-gameplay.
    - General bug with spectator screen.
- The screen won't exit automatically after completion.
    - Requires fixing spectator screen bug above + implementation.
- The local user state is set to `Idle` upon returning to the match subscreen. You have to click the spectate button again to spectate further.
    - Requires fixing spectator screen bug above + implementation.
- The quit button and other hud elements are visible but non-interactive.
    - Intended for now, HUDOverlay needs a pretty decent refactor.
- Spectating always starts from 0.
    - Needs a little bit more thinking as to the structure of things. Shouldn't be too difficult, but I want to keep things simple for this initial PR since it's already big.
    - Also requires fixes to spectator (https://github.com/ppy/osu/issues/12582).
- Lag with many players.
    - Yeah, 2000+ draw calls and 100K+ vertices drawn per frame is not easy to handle. Needs a lot of o!f support.

As for syncing, I'm taking a bit of a different approach than osu-stable for the time being, but I've added the appropriate abstractions to implement something similar to osu-stable if this doesn't work out. This implementation catches up clients that fall behind at 2x gameplay rate.
- Clients are kept within 50ms out of sync to the master clock.
- If a client desyncs, it catches up until it's within 16ms of the master clock. This probably doesn't need to be a different value than the above, but was required in the original implementation to avoid quick toggling of the catchup state.
- If a client desyncs into the future for any reason, it stops until the master clock catches up.

### CODE REVIEW

File: osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerSpectator.cs [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+            => AddAssert($"{userId} volume is {volume}", () => getInstance(userId).Volume.Value == volume);
+
+        private void waitForCatchup(int userId)
+            => AddUntilStep($"{userId} not catching up", () => !getInstance(userId).GameplayClock.IsCatchingUp);
+
+        private Player getPlayer(int userId) => getInstance(userId).ChildrenOfType<Player>().Single();
+
+        private PlayerArea getInstance(int userId) => spectatorScreen.ChildrenOfType<PlayerArea>().Single(p => p.UserId == userId);
+
+        public class TestSpectatorStreamingClient : SpectatorStreamingClient
```
Comment (smoogipoo) [inline osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerSpectator.cs:304]: This is effectively a duplicate of the one in a few other test scenes. It's my first priority to clean up after this is merged in - the implementation is slightly different in every case.

Please ignore for now, unless you really want this fixed here.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        private PlayerGrid grid;
+        private MultiSpectatorLeaderboard leaderboard;
+        private PlayerArea currentAudioSource;
+
+        /// <summary>
+        /// Creates a new <see cref="MultiSpectatorScreen"/>.
+        /// </summary>
+        /// <param name="userIds">The players to spectate.</param>
+        public MultiSpectatorScreen(int[] userIds)
+            : base(userIds.AsSpan().Slice(0, Math.Min(16, userIds.Length)).ToArray())
```
Comment (turbedi) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:45]:
```suggestion
            : base(userIds.AsSpan(0, Math.Min(16, userIds.Length)).ToArray())
```

File: osu.Game.Tests/OnlinePlay/TestCaseCatchUpSyncManager.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+using osu.Framework.Bindables;
+using osu.Framework.Testing;
+using osu.Framework.Timing;
+using osu.Game.Screens.OnlinePlay.Multiplayer.Spectate.Sync;
+using osu.Game.Tests.Visual;
+
+namespace osu.Game.Tests.OnlinePlay
+{
+    [HeadlessTest]
+    public class TestCaseCatchUpSyncManager : OsuTestScene
```
Comment (bdach) [inline osu.Game.Tests/OnlinePlay/TestCaseCatchUpSyncManager.cs:15]:
```suggestion
    [HeadlessTest]
    public class TestSceneCatchUpSyncManager : OsuTestScene
```

+ everything else that is needed (file rename at least)

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs (hunk 2 above)
Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:45]: At the point in which that span is terminated in a `.ToArray()` copy anyways I'd honestly be considering just doing `.Take(16)` for simplicity.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Container leaderboardContainer;
+            masterClockContainer = new MasterGameplayClockContainer(Beatmap.Value, 0);
+
+            InternalChildren = new[]
+            {
+                (Drawable)(syncManager = new CatchUpSyncManager(masterClockContainer)),
```
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:58]: Bit weird this, haven't seen that done much. Usually `Drawable` is hoisted out as the type of the array (so `new Drawable[] { ... }`)

**Review by bdach — COMMENTED** [review]
Had a rudimentary first pass, haven't run yet. Can't see anything major (yet?), just nits mostly.

**Comment by peppy** [comment]
With the listed known issues it sounds like this definitely isn't in a releasable state.. would be appreciated to know the path forward for this feature (how much more implementation work is required for a good user experience; if this is already partly done on another branch; if this is your next priority etc.)

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/Sync/ISlaveClock.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+using osu.Framework.Bindables;
+using osu.Framework.Timing;
+
+namespace osu.Game.Screens.OnlinePlay.Multiplayer.Spectate.Sync
+{
+    /// <summary>
+    /// A clock which is used by <see cref="MultiSpectatorPlayer"/>s and managed by an <see cref="ISyncManager"/>.
+    /// </summary>
+    public interface ISlaveClock : IFrameBasedClock, IAdjustableClock
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/Sync/ISlaveClock.cs:12]: disagree with any of the suggestions. slave is fine. this is code and "slave clock" is an industry wide term which perfectly explains the usage, unlike any of the alternatives.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:45]: Also probably worth mentioning this seemingly hidden limitation somewhere? Feels like hidden implementation that only 16 players are visible. Would want an inline `TODO` and/or xmldoc mention of this as a current(?) limitation.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs (hunk 4 above)
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:58]: As it is, it would give an R# inspection if I were to specify the array type. Would have to not use the interface and not cast to `Drawable` for that to work :(

**Comment by smoogipoo** [comment]
**tl;dr:** Anything but performance can probably be fixed in 1/2 weeks and fixing `SpectatorScreen` is my immediate priority.

> how much more implementation work is required for a good user experience

### `SpectatorScreen`

The issues of `SpectatorScreen` go quite deep, but is my immediate priority:
- `ScoreProcessor` doesn't have the correct `JudgedHits` state, leading to `HasCompleted` never turning true if spectated mid-gameplay.
- When a player stops playing, the spectator screen just resumes playback.
- Frames are added after being converted using the wrong (non-playable) beatmap.

Additionally, how does/will the above interact with the multiplayer gameplay leaderboard, which itself uses one `ScoreProcessor` for all players. Fixing some of this stuff may require server-side changes, that's why I'd call it pretty deep issues.

It could probably all be fixed within a week, just need to see what can be done and find a path forward.

### `MultiSpectatorScreen`

Spectating starts from 0 mostly because the gameplay start time stuff is immutable in `MasterGameplayClockContainer`, shouldn't take too long to fix.
The HUD, likewise, shouldn't take too long. I left it out because I think it's probably better implemented as something that you can derive instead, but it's very coupled to `Player` right now.

Also we discussed making the screen restart if the users start playing again afterwards, so I'll do that as a temporary thing.

### Performance

Definitely a many-month task:
- Most time is spent in draw, which is already hard to optimise.
- CursorTrail is a significant chunk of draw time.
- Pooling can be improved, but needs more support because everything's inside the playfields right now (or the drawable ruleset for samples).

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs (hunk 4 above)
Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:58]: Oh it's the compound assignment that does it... My bad, that's what I get for only reading the diff online without an IDE.

**Comment by peppy** [comment]
I know I already resolved the discussion on clock terminology, but coming from a different angle, I'm not sure the `master-slave` terms work here. Generally I would expect a master clock to not be changing its state based on the slaves' states, but in this scenario it looks to be only starting when slaves denote that they are ready.

Would you be against changing the master clock to something like `SyncMasterClock` (or just leaving it as it is, I guess) and slave to `ISpectatorPlayerClock`? I think this system is closed-loop enough to the spectator display that it can use more specific naming.

The thing that threw me off the most was the fact that the slave clocks are *passed in* to the slave for what that's worth. I almost thought you had things around back-to-front (may still suggest refactoring that to have the player areas create their own "slave" clock as I review the rest of this PR, if that seems feasible).

File: osu.Game.Tests/Visual/Multiplayer/TestSceneMultiSpectatorScreen.cs [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+            checkPausedInstant(56, true);
+
+            // Send frames for the other player, both should now start playing.
+            sendFrames(56, 20);
+            checkPausedInstant(55, false);
+            checkPausedInstant(56, false);
+        }
+
+        [Test]
+        public void TestPlayersDoNotStartSimultaneouslyIfBufferingForMaximumStartDelay()
```
Comment (peppy) [inline osu.Game.Tests/Visual/Multiplayer/TestSceneMultiSpectatorScreen.cs:132]: This test fails if run in isolation (have tested this was the case before my renaming changes):

20210426 164637 (JetBrains Rider)

**Comment by smoogipoo** [comment]
> Would you be against changing the master clock to something like SyncMasterClock (or just leaving it as it is, I guess) and slave to ISpectatorPlayerClock? I think this system is closed-loop enough to the spectator display that it can use more specific naming.

Sounds fine to me. What do you mean about `SyncMasterClock` though? Just the member name exposed from `ISyncManager`?

**Comment by peppy** [comment]
Yeah basically. I think just adding a `Clock` suffix to that member would be enough.

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs [hunk 7]
```diff
@@ -405,24 +406,45 @@ void endOperation()
...
 
-        private void onLoadRequested()
-        {
-            Debug.Assert(client.Room != null);
+            Debug.Assert(client.LocalUser != null);
 
-            int[] userIds = client.CurrentMatchPlayingUserIds.ToArray();
+            UpdateMods();
+
+            if (client.LocalUser.State == MultiplayerUserState.Spectating
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs:418]: Not a huge fan of this code path here. Feels like there already could be a race condition when changing states where `StartPlay` is called more than once.

Did you consider changing the `LoadRequested` event to also fire for spectators?

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        private readonly Bindable<bool> waitingOnFrames = new Bindable<bool>(true);
+        private readonly Score score;
+        private readonly ISpectatorPlayerClock spectatorPlayerClock;
+
+        /// <summary>
+        /// Creates a new <see cref="MultiSpectatorPlayer"/>.
+        /// </summary>
+        /// <param name="score">The score containing the player's replay.</param>
+        /// <param name="spectatorPlayerClock">The clock controlling the gameplay running state.</param>
+        public MultiSpectatorPlayer([NotNull] Score score, [NotNull] ISpectatorPlayerClock spectatorPlayerClock)
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: Wondering if we should be passing the master clock down to the player and having it create its client clock. The main reason I bring this up is the weirdness of binding to the property of something coming in via the `ctor` (and the fact that this is the only case of binding.. at which point maybe ownership makes more sense?).

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            spectatorPlayerClock.WaitingOnFrames.BindTo(waitingOnFrames);
+        }
+
+        protected override void UpdateAfterChildren()
+        {
+            base.UpdateAfterChildren();
+
+            // This is required because the frame stable clock is set to WaitingOnFrames = false for one frame.
+            waitingOnFrames.Value = DrawableRuleset.FrameStableClock.WaitingOnFrames.Value || score.Replay.Frames.Count == 0;
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:47]: Adding to above, especially with this setter here, I think I'd rather we were operating directly on the clock here (and potentially making the clock a nested class?)

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:58]: Shall we just change the field to the actual type rather than interface until we need it to be something else?

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/PlayerArea.cs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+using osu.Game.Scoring;
+using osu.Game.Screens.OnlinePlay.Multiplayer.Spectate.Sync;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Screens.OnlinePlay.Multiplayer.Spectate
+{
+    /// <summary>
+    /// Provides an area for and manages the hierarchy of a spectated player within a <see cref="MultiSpectatorScreen"/>.
+    /// </summary>
+    public class PlayerArea : CompositeDrawable, IAdjustableAudioComponent
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/PlayerArea.cs:26]: Rather than implementing `IAdjustableAudioComponent`, How about exposing a bool for the mute state? It seems like the only usage of this interface is setting `Volume` to 0 or 1.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs (hunk 8 above)
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: Players are late initialised, whenever they decide to say they've begun gameplay / that message is received from the spectator server. So I think this is the best that can be done, otherwise the sync manager will assume all players are loaded because their clocks haven't been added yet?

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs (hunk 9 above)
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:47]: As above, I can't really do this :/

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs (hunk 8 above)
Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: I see. Is that done with intent? ie. is there any reason we don't just load them all as soon as the screen loads then pause on the pre-gameplay phase (not saying we necessarily should, just curious why it is done this way).

Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: What do you load? Which beatmap/mods/ruleset? My primary thought going into it is that this supports that as is due to the nature of `SpectatorScreen`.

Otherwise, I could use the `PlaylistItem` for the beatmap/ruleset and extract the mods from each player (and leave the per-player beatmap/ruleset support until it's supported for multi in general), but then I also have to answer the question of how do I integrate that into `SpectatorScreen`? Sounds like at that point `SpectatorScreen` should basically not be overridden because it's a completely different flow of events.

Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: You're correct, but I don't think multiplayer should be a special case. That's an issue with spectator in general.

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: So would you see the data flow (as you described just above) being feasible to change at the `SpectatorScreen` level without too much fuss?

Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: Not the `SpectatorScreen` data flow. If multiplayer should be a special case then it basically cannot use `SpectatorScreen` due to the complexity.

The problem is the clients being spectated only send their gameplay started event in `LoadComplete`:
https://github.com/ppy/osu/blob/72d9cdcdf919a9ed4032eeb99742e5783e5781d4/osu.Game/Rulesets/UI/ReplayRecorder.cs#L46-L53

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: Hmm.. so you're saying the way forward you see is to start loading the player when they player starts loading, rather than when they start sending frames? And making spectator also do that?

I'm mostly thinking about how we would want a tournament stream to look going forward, and whether this aligns with that vision, if that gives context to why I'm bringing this up.

Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: As for tournaments, wouldn't it make more sense to always start gameplay at 0? At which point this becomes a problem of making sure that we've received all earlier frames (which I'm pretty sure doesn't hapen if we start spectating during gameplay, for example).

> Hmm.. so you're saying the way forward you see is to start loading the player when they player starts loading, rather than when they start sending frames? And making spectator also do that?

Exactly.

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: I'm going to need a bit of time to think about this. I was hoping we'd be able to provide a slightly less "flashy" view in the case where all clients are loaded and playing, but maybe as you say is fine.

As for "always starting from 0", we also need to consider the case of starting spectating from mid-game, which should definitely be possible (consider a scenario where lazer crashes for whatever reason and you want to pick back up on watching an ongoing match).

Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: Maybe it's a matter of providing a button to go to the current time? I could imagine this being useful in normal spectator too, with its rewind functionality.

Comment (peppy) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs:29]: Probably should default to current time, but sure.

Also brings up a point of when seeking support will be added to this screen (should be added soon after implementation as it feels like one of the minimum requirements for the feature).

File: osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs (hunk 7 above)
Comment (smoogipoo) [reply osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs]: Sure, I've applied the necessary changes to make this work, requires server-side change.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        private PlayerArea currentAudioSource;
+
+        /// <summary>
+        /// Creates a new <see cref="MultiSpectatorScreen"/>.
+        /// </summary>
+        /// <param name="userIds">The players to spectate.</param>
+        public MultiSpectatorScreen(int[] userIds)
+            : base(userIds.Take(PlayerGrid.MAX_PLAYERS).ToArray())
+        {
+            this.userIds = GetUserIds().ToArray();
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:53]: This is a mouthful....

File: osu.Game/Screens/Spectate/SpectatorScreen.cs [hunk 12]
```diff
@@ -26,7 +27,8 @@ namespace osu.Game.Screens.Spectate
     /// </summary>
     public abstract class SpectatorScreen : OsuScreen
     {
-        private readonly int[] userIds;
+        protected IEnumerable<int> GetUserIds() => userIds;
+        private readonly HashSet<int> userIds = new HashSet<int>();
```
Comment (peppy) [inline osu.Game/Screens/Spectate/SpectatorScreen.cs:31]: Rather than the weird `GetUserIds()` method, are you okay with making this a `List<>` and exposing as a `protected IReadOnlyList<>` for use in `MultiplayerSpectatorScreen`?

Comment (peppy) [reply osu.Game/Screens/Spectate/SpectatorScreen.cs:31]: I've made this change anyways, see if you agree with it.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/Sync/CatchUpSyncManager.cs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+                }
+
+                // Make sure the player clock is running if it can.
+                if (!clock.WaitingOnFrames.Value)
+                    clock.Start();
+
+                if (clock.IsCatchingUp)
+                {
+                    // Stop the player clock from catching up if it's within the sync target.
+                    if (timeDelta <= SYNC_TARGET)
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/Sync/CatchUpSyncManager.cs]: I think we're going to want to set this to `SYNC_TARGET / 2`, else there may be a case of the state flip-flopping between catch-up and non-catchup.

**Review by peppy — DISMISSED** [review]
I think this is in a good state to get it merged for further testing / improvement.

@smoogipoo please go through the changes I made and make sure you're okay with them.

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
This hard crashes if you spectate a room while not having the beatmap locally.

```csharp

[runtime] 2021-05-03 07:55:04 [verbose]: Screen changed → MultiSpectatorScreen
[runtime] 2021-05-03 07:55:04 [error]: An unhandled error has occurred.
[runtime] 2021-05-03 07:55:04 [error]: System.InvalidOperationException: Sequence contains no elements
[runtime] 2021-05-03 07:55:04 [error]: at System.Linq.ThrowHelper.ThrowNoElementsException()
[runtime] 2021-05-03 07:55:04 [error]: at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
[runtime] 2021-05-03 07:55:04 [error]: at osu.Game.Screens.Play.MasterGameplayClockContainer..ctor(WorkingBeatmap beatmap, Double gameplayStartTime, Boolean startAtGameplayStart) in /Users/dean/Projects/osu/osu.Game/Screens/Play/MasterGameplayClockContainer.cs:line 68
[runtime] 2021-05-03 07:55:04 [error]: at osu.Game.Screens.OnlinePlay.Multiplayer.Spectate.MultiSpectatorScreen.load() in /Users/dean/Projects/osu/osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs:line 57
[runtime] 2021-05-03 07:55:04 [error]: --- End of stack trace from previous location ---
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Drawable.LoadFromAsync(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.loadComponents[TLoadable](List`1 components, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext, CancellationToken cancellation)
[runtime] 2021-05-03 07:55:04 [error]: at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
[runtime] 2021-05-03 07:55:04 [error]: --- End of stack trace from previous location ---
[runtime] 2021-05-03 07:55:04 [error]: at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
[runtime] 2021-05-03 07:55:04 [error]: --- End of stack trace from previous location ---
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.<>c__DisplayClass17_2`1.<LoadComponentsAsync>b__4()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Threading.Scheduler.Update()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2021-05-03 07:55:04 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
```

**Comment by frenzibyte** [comment]
FWIW this doesn't guard against the host switching to another beatmap while in spectating mode, but I guess that can come as a follow-up PR.

**Comment by smoogipoo** [comment]
> FWIW this doesn't guard against the host switching to another beatmap while in spectating mode, but I guess that can come as a follow-up PR.

I don't understand what you mean. They are supposed to be able to change the beatmap - they're the host after all.

Unless you mean that they're able to open song select after exiting gameplay. This is a thing for normal hosts too, but the actual changing of the beatmap is blocked server-side. The button should probably be disabled, but yes that's for a separate PR.

**Comment by frenzibyte** [comment]
Last time I checked, I was still in spectating mode while the user has switched to another beatmap I don't have, will double-check though as I may potentially be wrong.

**Comment by smoogipoo** [comment]
Right, then your issue is the same as above, perhaps.

**Comment by smoogipoo** [comment]
> This hard crashes if you spectate a room while not having the beatmap locally.

I've fixed this one above.

I've also grouped in a change which fixes the case when the user is switched from ready -> idle upon deleting a beatmap. It was comparing objects via `!=` rather than the availability state or using `.Equals()`:
https://github.com/ppy/osu/blob/4be15cfc5a47a839ebd303cbaa3a7795759191d3/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs#L356-L359

An appropriate test has also been added for all of this.

> And if you have the beatmap available:

As for this, I don't know how to repro this, but I can _imagine_ how it could happen. Was it just a random failure when everything else was in a good state?

File: osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerSpectator.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerSpectator.cs]: This has now been removed, and all implementations combined as of https://github.com/ppy/osu/pull/12538/commits/21fc0ba43baea6e0cb115952ab21df06f1d1c003

**Comment by smoogipoo** [comment]
I've resolved the above crash as well - it's just due to the server sending `UserBeganPlaying()` regardless of whether the users are being watched. I didn't realise this was a thing when implementing `SpectatorScreen`.

File: osu.Game/Tests/Visual/Spectator/TestSpectatorStreamingClient.cs [hunk 14]
```diff
@@ -14,7 +15,7 @@ namespace osu.Game.Tests.Visual.Spectator
     public class TestSpectatorStreamingClient : SpectatorStreamingClient
     {
         public new BindableList<int> PlayingUsers => (BindableList<int>)base.PlayingUsers;
-        private readonly HashSet<int> watchingUsers = new HashSet<int>();
+        private readonly ConcurrentDictionary<int, byte> watchingUsers = new ConcurrentDictionary<int, byte>();
```
Comment (peppy) [inline osu.Game/Tests/Visual/Spectator/TestSpectatorStreamingClient.cs:18]: lol what is this. just working around lack of `ConcurrentList`? does `ConcurrentBag` not work here?

Comment (smoogipoo) [reply osu.Game/Tests/Visual/Spectator/TestSpectatorStreamingClient.cs:18]: ConcurrentBag doesn't have a `TryAdd`, which is what I wanted in particular. I can change it to use locking as an alternative.

Comment (peppy) [reply osu.Game/Tests/Visual/Spectator/TestSpectatorStreamingClient.cs:18]: I think it's fine considering it's just a test. Maybe a comment.

**Comment by peppy** [comment]
Let's get this merged. There's still a few things left to tidy up before this feels like it's ready for users' hands (fixing transparent backgrounds showing other clients through them; seeking forward to earliest spectator frames when watching an ongoing play are the two that i noticed) but this PR is already too old/large!

**Review by peppy — APPROVED** [review]
